bokehjs/src/lib/core/kinds.ts

```diff
@@ -0,0 +1,218 @@
+import * as tp from "./util/types"
+import {Color as ColorType} from "./types"
+import {is_color} from "./util/color"
+
+export abstract class Kind<T> {
+  __type__: T
+
+  abstract valid(value: unknown): value is this["__type__"]
+}
+
+type Constructor<T> = Function & {prototype: T}
+
+export namespace Kinds {
+  export class Any extends Kind<any> {
+    valid(_value: unknown): _value is any {
+      return true
+    }
+  }
+
+  export class Unknown extends Kind<unknown> {
+    valid(_value: unknown): _value is unknown {
+      return true
+    }
+  }
+
+  export class Boolean extends Kind<boolean> {
+    valid(value: unknown): value is boolean {
+      return tp.isBoolean(value)
+    }
+  }
+
+  export class Ref<ObjType extends object> extends Kind<ObjType> {
+    constructor(readonly obj_type: Constructor<ObjType>) {
+      super()
+    }
+
+    valid(value: unknown): value is ObjType {
+      // XXX: disable validation for now, because object graph
+      // initialization depends on this.
+      value
```

Comment:
@mattpap why add this useless expression here instead of `_value` on line 37?

Reply:
This because `value` is used in the return type, however due to TypeScript's limitation (bug?), it doesn't register this as a usage, so would complain about an unused variable. In any case, I will be fixing the underlying issue soon.